### The Problem Statement
Your company is expanding in to new industries to diversify its portfolio. Specifically, they are interested in purchasing and operating airplanes for commercial and private enterprises, but do not know anything about the potential risks of aircraft. You are charged with determining which aircraft are the lowest risk for the company to start this new business endeavor. You must then translate your findings into actionable insights that the head of the new aviation division can use to help decide which aircraft to purchase.

### The Business Questions
1. What are the safety records and incident rates for different types of aircraft? (aircraft models, incident type, failure points - engine, structure, weather, location)

2. How do different aircraft perform under varying operational environments including weather patterns?
  
3. What is the relation between the quality of aircraft and safety in relation to make, model, engine type and number of engines?

### Objectives 
- Identify aircraft models with the lowest incident rates and the factors contributing to their safety performance.

- Examine the impact of weather conditions on different aircraft models.

- Investigate how the make, model, engine type, and number of engines correlate with safety performance and reliability.

In understanding the `AviationData.csv`, necessary libraries are to be imported for proper data analysis and visualization.

In [72]:
# Importing necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

The `AviationData.csv` file is then loaded to `aviation_df` through Pandas. The `encoding` parameter `latin-1` was called to ensure decoding of the file is done correctly to avoid crashing. The `low_memory` parameter `False` was called to be able to optimize memory usage and ensure consistent column data types.

In [73]:
# Loading the data to aviation_df and calling the first 5 rows
aviation_df = pd.read_csv('AviationData.csv', encoding='latin-1', low_memory=False)
aviation_df.head()

Event.Id Investigation.Type Accident.Number  Event.Date  \
0  20001218X45444           Accident      SEA87LA080  1948-10-24   
1  20001218X45447           Accident      LAX94LA336  1962-07-19   
2  20061025X01555           Accident      NYC07LA005  1974-08-30   
3  20001218X45448           Accident      LAX96LA321  1977-06-19   
4  20041105X01764           Accident      CHI79FA064  1979-08-02   

          Location        Country   Latitude   Longitude Airport.Code  \
0  MOOSE CREEK, ID  United States        NaN         NaN          NaN   
1   BRIDGEPORT, CA  United States        NaN         NaN          NaN   
2    Saltville, VA  United States  36.922223  -81.878056          NaN   
3       EUREKA, CA  United States        NaN         NaN          NaN   
4       Canton, OH  United States        NaN         NaN          NaN   

  Airport.Name  ... Purpose.of.flight Air.carrier Total.Fatal.Injuries  \
0          NaN  ...          Personal         NaN                  2.0   
1          NaN  ...          Personal         NaN                  4.0   
2          NaN  ...          Personal         NaN                  3.0   
3          NaN  ...          Personal         NaN                  2.0   
4          NaN  ...          Personal         NaN                  1.0   

  Total.Serious.Injuries Total.Minor.Injuries Total.Uninjured  \
0                    0.0                  0.0             0.0   
1                    0.0                  0.0             0.0   
2                    NaN                  NaN             NaN   
3                    0.0                  0.0             0.0   
4                    2.0                  NaN             0.0   

  Weather.Condition  Broad.phase.of.flight   Report.Status Publication.Date  
0               UNK                 Cruise  Probable Cause              NaN  
1               UNK                Unknown  Probable Cause       19-09-1996  
2               IMC                 Cruise  Probable Cause       26-02-2007  
3               IMC                 Cruise  Probable Cause       12-09-2000  
4               VMC               Approach  Probable Cause       16-04-1980  

[5 rows x 31 columns]

In understanding the dataframe, `df.info()` is run to get the columns names, total entries, any missing values in the columns and the data types of the columns. Afterwards, `df.duplicated()` is run to find any duplicated values if any. `df.isna().sum()` is the applied to be able to find the count of missing values in the columns. 

In [ ]:
# Inspecting the information of aviation_df
aviation_df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 31 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Latitude                34382 non-null  object 
 7   Longitude               34373 non-null  object 
 8   Airport.Code            50249 non-null  object 
 9   Airport.Name            52790 non-null  object 
 10  Injury.Severity         87889 non-null  object 
 11  Aircraft.damage         85695 non-null  object 
 12  Aircraft.Category       32287 non-null  object 
 13  Registration.Number     87572 non-null  object 
 14  Make                    88826 non-null

In [75]:
# Finding the shape of the dataframe
aviation_df.shape

(88889, 31)

In [76]:
aviation_df.nunique() # Finding the number of unique values in each column

Event.Id                  87951
Investigation.Type            2
Accident.Number           88863
Event.Date                14782
Location                  27758
Country                     219
Latitude                  25589
Longitude                 27154
Airport.Code              10375
Airport.Name              24871
Injury.Severity             109
Aircraft.damage               4
Aircraft.Category            15
Registration.Number       79105
Make                       8237
Model                     12318
Amateur.Built                 2
Number.of.Engines             7
Engine.Type                  13
FAR.Description              31
Schedule                      3
Purpose.of.flight            26
Air.carrier               13590
Total.Fatal.Injuries        125
Total.Serious.Injuries       50
Total.Minor.Injuries         57
Total.Uninjured             379
Weather.Condition             4
Broad.phase.of.flight        12
Report.Status             17075
Publication.Date           2924
dtype: i

In [77]:
# Inspecting for duplicated values in the dataframe
aviation_df.duplicated().sum()

0

In [ ]:
# Inspecting missing values from aviation_df 
aviation_df.isna().sum() 

Event.Id                      0
Investigation.Type            0
Accident.Number               0
Event.Date                    0
Location                     52
Country                     226
Latitude                  54507
Longitude                 54516
Airport.Code              38640
Airport.Name              36099
Injury.Severity            1000
Aircraft.damage            3194
Aircraft.Category         56602
Registration.Number        1317
Make                         63
Model                        92
Amateur.Built               102
Number.of.Engines          6084
Engine.Type                7077
FAR.Description           56866
Schedule                  76307
Purpose.of.flight          6192
Air.carrier               72241
Total.Fatal.Injuries      11401
Total.Serious.Injuries    12510
Total.Minor.Injuries      11933
Total.Uninjured            5912
Weather.Condition          4492
Broad.phase.of.flight     27165
Report.Status              6381
Publication.Date          13771
dtype: i

In [79]:
# Checking the names of the dataframe's columns
aviation_df.columns

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Latitude', 'Longitude', 'Airport.Code',
       'Airport.Name', 'Injury.Severity', 'Aircraft.damage',
       'Aircraft.Category', 'Registration.Number', 'Make', 'Model',
       'Amateur.Built', 'Number.of.Engines', 'Engine.Type', 'FAR.Description',
       'Schedule', 'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date'],
      dtype='object')

#### Summary of the Dataset
- The dataset has 31 columns and 88889 rows. 
- The datatypes include objects and floats. 
- Just 4 columns do not have missing values i.e. `Event.Id`, `Investigation.Type`, `Accident.Number` and `Event.Date`

### Cleaning and Normalizing the Data

After identyfing the dataframe's information, shape, the names of the columns, the summary statistics of numerical columns, the data cleaning process commences. First, there are certain columns with many missing values and may not be of much help eventually in the analysis. These include `Latitude` and `Longitude` columns. 

In [81]:
# Dropping the 'Latitude' and 'Longitude' columns
aviation_df = aviation_df.drop(['Latitude', 'Longitude'], axis=1)

In [82]:
# Checking to see if the columns have been dropped.
aviation_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 88889 entries, 0 to 88888
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Event.Id                88889 non-null  object 
 1   Investigation.Type      88889 non-null  object 
 2   Accident.Number         88889 non-null  object 
 3   Event.Date              88889 non-null  object 
 4   Location                88837 non-null  object 
 5   Country                 88663 non-null  object 
 6   Airport.Code            50249 non-null  object 
 7   Airport.Name            52790 non-null  object 
 8   Injury.Severity         87889 non-null  object 
 9   Aircraft.damage         85695 non-null  object 
 10  Aircraft.Category       32287 non-null  object 
 11  Registration.Number     87572 non-null  object 
 12  Make                    88826 non-null  object 
 13  Model                   88797 non-null  object 
 14  Amateur.Built           88787 non-null

In [83]:
aviation_df.describe() # Summary statistics of numerical columns

Number.of.Engines  Total.Fatal.Injuries  Total.Serious.Injuries  \
count       82805.000000          77488.000000            76379.000000   
mean            1.146585              0.647855                0.279881   
std             0.446510              5.485960                1.544084   
min             0.000000              0.000000                0.000000   
25%             1.000000              0.000000                0.000000   
50%             1.000000              0.000000                0.000000   
75%             1.000000              0.000000                0.000000   
max             8.000000            349.000000              161.000000   

       Total.Minor.Injuries  Total.Uninjured  
count          76956.000000     82977.000000  
mean               0.357061         5.325440  
std                2.235625        27.913634  
min                0.000000         0.000000  
25%                0.000000         0.000000  
50%                0.000000         1.000000  
75%                0.000000         2.000000  
max              380.000000       699.000000

The aviation industry underwent significant changes in the 1980s, including advancements in technology, stricter safety regulations and improved reporting systems. This therefore warranted dropping data earlier than 1982. 

In [84]:
aviation_df = aviation_df[aviation_df['Event.Date'] >= '1982-01-01']

Among the columns is `Event.Date` column that its data type is an object. To be able to find the year and months of the event, best case is to convert the `Event.Date` column to a Datetime data type. There after, add new columns of `Month` and `Year`.  

In [85]:
# Changing the 'Event.Date' to datetime
aviation_df['Event.Date'] = pd.to_datetime(aviation_df['Event.Date'])
aviation_df.info() # To check the dtype of 'Event.Date' column

<class 'pandas.core.frame.DataFrame'>
Int64Index: 88882 entries, 7 to 88888
Data columns (total 29 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Event.Id                88882 non-null  object        
 1   Investigation.Type      88882 non-null  object        
 2   Accident.Number         88882 non-null  object        
 3   Event.Date              88882 non-null  datetime64[ns]
 4   Location                88830 non-null  object        
 5   Country                 88656 non-null  object        
 6   Airport.Code            50249 non-null  object        
 7   Airport.Name            52790 non-null  object        
 8   Injury.Severity         87882 non-null  object        
 9   Aircraft.damage         85688 non-null  object        
 10  Aircraft.Category       32286 non-null  object        
 11  Registration.Number     87565 non-null  object        
 12  Make                    88819 non-null  object

In [86]:
# Adding the 'Year' and 'Month' columns to the aviation_df
aviation_df['Year'] = aviation_df['Event.Date'].dt.year
aviation_df['Month'] = aviation_df['Event.Date'].dt.month_name()

In [87]:
aviation_df['Year'].head() # Checking to see if the 'Year' column has been added


7     1982
8     1982
9     1982
10    1982
11    1982
Name: Year, dtype: int64

In [88]:
aviation_df['Month'].head() # Checking to see if the 'Month' column has been added

7     January
8     January
9     January
10    January
11    January
Name: Month, dtype: object

In [89]:
# Adding 'Total_Injuries' column to sum all the injured i.e. 'Total.Fatal.Injuries', 'Total.Serious.Injuries' and 'Total.Minor.Injuries'
aviation_df['Total.Injuries'] = aviation_df['Total.Fatal.Injuries'] + aviation_df['Total.Serious.Injuries'] + aviation_df['Total.Minor.Injuries']
aviation_df.columns # Checking to see if the column has been added

Index(['Event.Id', 'Investigation.Type', 'Accident.Number', 'Event.Date',
       'Location', 'Country', 'Airport.Code', 'Airport.Name',
       'Injury.Severity', 'Aircraft.damage', 'Aircraft.Category',
       'Registration.Number', 'Make', 'Model', 'Amateur.Built',
       'Number.of.Engines', 'Engine.Type', 'FAR.Description', 'Schedule',
       'Purpose.of.flight', 'Air.carrier', 'Total.Fatal.Injuries',
       'Total.Serious.Injuries', 'Total.Minor.Injuries', 'Total.Uninjured',
       'Weather.Condition', 'Broad.phase.of.flight', 'Report.Status',
       'Publication.Date', 'Year', 'Month', 'Total.Injuries'],
      dtype='object')